cmd/snap-update-ns/system.go

```diff
@@ -66,9 +66,13 @@ func (upCtx *SystemProfileUpdateContext) Assumptions() *Assumptions {
 	//
 	// /snap/$SNAP_INSTANCE_NAME and /snap/$SNAP_NAME are added to allow
 	// remapping for parallel installs only when the snap has an instance key
+	//
+	// /var/lib/snapd/hostfs/var is there to allow the content interface to
+	// bridge to arbitrary sub-directories of /var from the host, even if they
+	// don't exist (they are created on the host in such case).
 	as := &Assumptions{}
 	instanceName := upCtx.InstanceName()
-	as.AddUnrestrictedPaths("/tmp", "/var/snap", "/snap/"+instanceName)
+	as.AddUnrestrictedPaths("/tmp", "/var/snap", "/snap/"+instanceName, "/var/lib/snapd/hostfs/var")
```

Comment:
what's the interaction if any of this change with layouts (vs mount by interfaces) ?

Reply:
layouts and mount by interfaces use the same mechanics but there are some important differences. First of all, layouts are defined by a boolean flag mentioned in the mount options list. The mount profile is then applied by snap-update-ns and layouts have a small difference in as to which directories are created (layouts create both source and target, normal mounts only create the target). Finally we come to the interaction with this list.

Normally mkdir is not even attempted (note that we don't check if a path component exists ahead of time, we try to create it from scratch, we do check if the entire path exists, this is important below) if the parent directory belongs to a writable filesystem. Read-only filesystems are by definition read only so writing to them is not restricted. Writing to a writable directory that is not on the list extended by this patch, causes a mimic to be constructed, so that the created directory is ephemeral. Therefore extending this list allows layouts and content interfaces, if they are defined to use `/var/lib/snapd/hostfs/var`, to create placeholder files and directories that are visible in the host.

The only other consumer of mount entries that uses hostfs is the desktop interface. It works with a list of "font directories", checking if each entry exists (this is important, the reference above is explained in a moment) and if so, uses hostfs to bring it into the mount namespace.

Now for the important connection. We don't attempt to create the non-layout mount source (only mount target) so this extension to unrestricted paths does not interact with any other interface as the only possible candidate, desktop, uses the optimized path because both snapd and snap-update-ns check if the font directory fully exists ahead of considering it.

Reply:
are you saying that with that change a layout can now create directories under /var in the host?

Reply:
Layouts cannot create anything in /var/lib/snapd/*, which includes hostfs, it would allow a snap to create a directory on the host if that location is bridged with snap-confine's default set - this includes /var/snap (but that is black-listed from layouts), /var/lib/snapd (also black-listed), /var/tmp (allowed), /var/log (allowed) and finally /var/lib/extrausers (allowed)

Reply:
> pedronis: (to be precise, users cannot request a layout, not that layouts cannot create)
> 17:44 pedronis: interestingly, /var/log snap-confine-made, fixed mount, is marked with a TODO, asking to move it to an interface
> 17:44 pedronis: I think the approach is right but perhaps we need to investigate the unrestricted path more 
> 17:44 pedronis: here we _do_ want it (in /var/lib/dhcp) because otherwise we'll end up with a mimic
> 17:44 pedronis: but the point of the interface is to expose _real_ /var/lib/dhcp to the snap, creating it if necessary
> 17:45 pedronis: I would be happy with a special case that says var/lib/dhcp can be made
> 17:45 and continue the investigation into what is exactly allowed per interface, akin to what Maciek hinted at
> 17:45 
> <pedronis> Samuele Pedroni zyga: to be clear I'm slightly less worried about interfaces, my worry is layout, whether we have enough checks in place, not to make something odd happens
> 17:45 if we change something there
> 17:45 because the new unrestriction
> 17:46 
> <zyga> Zygmunt Bazyli Krynicki pedronis: so, users cannot request a layout to /var/lib/snapd/hostfs/* so the answer is that it is not something that interacts with layouts directly
> 17:47 pedronis: and actually, thinking about it now
> 17:48 pedronis: my comment was incorrect - given that layouts and snap-confine made mount points don't interact
> 17:48 pedronis: this doesn't change anything layouts can make
> 17:48 pedronis: I was confused because if you put a layout from $SNAP/foo to /var/lib/foo 
> 17:48 that feels like it might interact
> 17:48 
> <zyga> Zygmunt Bazyli Krynicki because perhaps /var/lib/foo is bridged by snap-confine to the host
> 17:48 but that's irrelevant for /var/lib/snapd/hostfs/var/ that is allowed
> 17:49 because /var/lib/foo is not in a prefix of hostfs, you end up with a mimic
> 17:49 as such I think this is safer than I assumed, since nothing apart from snapd code can request new hostfs entries